src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala

```diff
@@ -246,40 +247,63 @@ trait WorkspaceComponent {
-      val workflowDatesQuery = for {
-        submissions <- submissionQuery if submissions.workspaceId.inSetBind(workspaceIds)
-        workflows <- workflowQuery if submissions.id === workflows.submissionId
-      } yield (submissions.workspaceId, workflows.status, workflows.statusLastChangedDate)
 
-      val workflowDatesGroupedQuery = workflowDatesQuery.groupBy { case (wsId, status, _) => (wsId, status) }.
-        map { case ((wsId, wfStatus), records) => (wsId, wfStatus, records.map { case (_, _, lastChanged) => lastChanged }.max) }
-
-      // running submission query: select workspaceId, count(1) ... where submissions.status === Submitted group by workspaceId
-      val runningSubmissionsQuery = (for {
-        submissions <- submissionQuery if submissions.workspaceId.inSetBind(workspaceIds) && submissions.status === SubmissionStatuses.Submitted.toString
-      } yield submissions).groupBy(_.workspaceId).map { case (wfId, submissions) => (wfId, submissions.length)}
-
-      for {
-        workflowDates <- workflowDatesGroupedQuery.result
-        runningSubmissions <- runningSubmissionsQuery.result
-      } yield {
-        val workflowDatesByWorkspaceByStatus: Map[UUID, Map[String, Option[Timestamp]]] = groupByWorkspaceIdThenStatus(workflowDates)
-        val runningSubmissionCountByWorkspace: Map[UUID, Int] = groupByWorkspaceId(runningSubmissions)
-
-        workspaceIds.map { wsId =>
-          val (lastFailedDate, lastSuccessDate) = workflowDatesByWorkspaceByStatus.get(wsId) match {
-            case None => (None, None)
-            case Some(datesByStatus) =>
-              (datesByStatus.getOrElse(WorkflowStatuses.Failed.toString, None), datesByStatus.getOrElse(WorkflowStatuses.Succeeded.toString, None))
-          }
-          wsId -> WorkspaceSubmissionStats(
-            lastSuccessDate.map(t => new DateTime(t.getTime)),
-            lastFailedDate.map(t => new DateTime(t.getTime)),
-            runningSubmissionCountByWorkspace.getOrElse(wsId, 0)
-          )
+      case class StatRow(
+        workspaceId: UUID,
+        cnt: Int,
+        dt: DateTime,
+        status: String)
+
+      implicit val queryToStatRow = GetResult(r => StatRow(r.<<, r.<<, r.<<, r.<<))
+
+      // by using a literal string inside the where clause, we are technically vulnerable to SQL injection.
+      // however, the UUIDs sent in to the literal are generated by the database; the user has no
+      // direct access to this argument
+      val inUUIDs = workspaceIds.map{u => "X'" + u.toString.replace("-","") + "'"}.mkString(",")
+
+      // we abuse UNION here to optimize database interaction. We make one query to return both the
+      // count-of-running-submissions and the last-succeeded/last-failed workflows.
+      // the 'cnt' column will be -1 for workflows, and a value for submissions.
+      import driver.quoteIdentifier
+      val sql:ReadAction[Seq[StatRow]] =
```

Comment:
I'll investigate this once I get back to a dev environment. I thought I saw that Slick generated poor queries for this, but I don't have the info handy and could be remembering wrong.